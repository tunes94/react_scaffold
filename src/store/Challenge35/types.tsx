import * as React from "react";

export enum Actions {
  ADD_TODO = "ADD_TODO",
  EDIT_TODO = "EDIT_TODO",
  DELETE_TODO = "DELETE_TODO"
}
