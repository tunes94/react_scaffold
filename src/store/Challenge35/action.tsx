import * as React from "react";
import { number, string } from "prop-types";
import { Actions } from "./types";
import { action, EmptyAction } from "typesafe-actions";

export const addToDo = (text: string): EmptyAction<string> => {
  return action(Actions.ADD_TODO, { text });
};

// export function addToDo(text: string): any {
//   return {
//     type: Actions.ADD_TODO,
//     payload: {
//       text
//     }
//   };
// }
