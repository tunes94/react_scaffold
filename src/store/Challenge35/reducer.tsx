import { Actions } from "./types";

export interface ToDo {
  text: string;
  todo_state: boolean;
  todo_id: number;
}

// interface initiaLStateInteface {
//   toDos: ToDo[];
// }

// const INITIAL_STATE: initiaLStateInteface = {
//   toDos: []
// };

const INITIAL_STATE: {
  toDos: ToDo[];
} = {
  toDos: []
};

export default function challenge35Reducer(state = INITIAL_STATE, action: any) {
  switch (action.type) {
    case Actions.ADD_TODO: {
      // throw {};
      return {
        toDos: [
          ...state.toDos,
          {
            text: action.payload.text,
            todo_state: false,
            todo_id: Math.floor(Math.random() * 1000)
          }
        ]
      };
    }
    case Actions.EDIT_TODO: {
      let idSelected = action.payload.todo_id;
      let toDosCopy: ToDo[] = [...state.toDos];

      toDosCopy.forEach((toDo: ToDo) => {
        if (idSelected === toDo.todo_id) {
          toDo.todo_state = !toDo.todo_state;
        }
        return toDo;
      });
      return {
        toDos: toDosCopy
      };
    }
    case Actions.DELETE_TODO: {
      let idSelected = action.payload.todo_id;
      let toDosCopy: ToDo[] = [...state.toDos];
      let filteredToDos: ToDo | any = toDosCopy.filter(
        todo => todo.todo_id !== idSelected
      );
      console.log(filteredToDos);

      return { toDos: filteredToDos };
    }
    default:
      return state;
  }
}
