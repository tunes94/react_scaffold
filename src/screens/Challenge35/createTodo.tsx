import * as React from "react";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { addToDo } from "../../store/Challenge35/action";

export interface CreateToDoProps {
  //   text: string;
  addToDo?: (text: string) => void;
}

export interface InternalState {
  text: string;
}

class CreateToDo extends React.Component<CreateToDoProps, InternalState> {
  constructor(props: CreateToDoProps) {
    super(props);
    this.state = {
      text: ""
    };
  }

  handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({
      text: event.target.value
    });
  };

  render(): JSX.Element {
    const { text } = this.state;
    return (
      <div className="container">
        <h1>Challenge 3.5</h1>
        <div className="container shadow">
          <div className="form-inline p-3">
            <h5>Create a new To Do</h5>
            <input
              name="text"
              value={text}
              onChange={this.handleChange}
              type="text"
              className="form-control col-6 m-3"
              aria-describedby="helpId"
              placeholder="To Do text"
            ></input>

            <button
              onClick={(): void => {
                try {
                  this.props.addToDo && this.props.addToDo(text);
                  alert("New To Do created with success!!");
                } catch {
                  alert("Error!");
                }
              }}
              type="button"
              className="btn btn-outline-dark"
            >
              Submit
            </button>
          </div>
        </div>
      </div>
    );
  }
}
const mapDispatchToProps = (dispatch: any): any =>
  bindActionCreators({ addToDo }, dispatch);

export default connect(
  null,
  mapDispatchToProps
)(CreateToDo);
