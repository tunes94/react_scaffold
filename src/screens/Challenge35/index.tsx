import React from "react";
import { connect } from "react-redux";
import { ToDo } from "../../store/Challenge35/reducer";
import { Table } from "react-bootstrap";
import { bindActionCreators } from "redux";
import { editToDo, deleteToDo } from "../../store/Challenge35/action";

export interface Challenge35Props {
  toDos: ToDo[];
  editToDo?: (todo_id: number) => void;
  deleteToDo?: (todo_id: number) => void;
  // text: string;
  // addToDo?: (text: string) => void;
}

export interface InternalState {
  // text: string;
}

class Challenge35 extends React.Component<Challenge35Props, InternalState> {
  constructor(props: Challenge35Props) {
    super(props);
    this.state = {
      // text: ""
    };
  }

  public render(): JSX.Element {
    const { toDos, editToDo, deleteToDo } = this.props;
    return (
      <div className="container">
        <h1>Challenge 3.5</h1>
        {toDos.length ? (
          <div className="container shadow mt-5">
            <h4>List of To Do's</h4>
            <Table style={{ marginTop: "50px" }} striped bordered hover>
              <thead>
                <tr>
                  <th scope="col">ID</th>
                  <th scope="col">State</th>
                  <th scope="col">Text</th>
                  <th scope="col">Operations</th>
                </tr>
              </thead>

              <tbody>
                {toDos &&
                  toDos.map((todo: ToDo) => (
                    <tr key={todo.todo_id}>
                      <td>{todo.todo_id}</td>
                      <td>{todo.todo_state ? "Done" : "To Do"}</td>
                      <td>{todo.text}</td>

                      <td>
                        <button
                          onClick={(): void => {
                            try {
                              editToDo && editToDo(todo.todo_id);
                            } catch {
                              alert("Error!");
                            }
                          }}
                          className="btn btn-dark col-3 ml-5"
                        >
                          Change State
                        </button>

                        <button
                          onClick={(): void => {
                            if (todo.todo_state === true) {
                              deleteToDo && deleteToDo(todo.todo_id);
                            } else {
                              alert("Complete this To Do before deleting it!");
                            }
                          }}
                          className="btn btn-danger col-2 ml-5"
                        >
                          Delete
                        </button>
                      </td>
                    </tr>
                  ))}
              </tbody>
            </Table>
          </div>
        ) : (
          <h5 className="mt-5">
            There are no To Do's to show at the moment...
          </h5>
        )}
      </div>
    );
  }
}

const mapStateToProps = (state?: any) => ({
  toDos: state.challenge35Reducer.toDos
});

const mapDispatchToProps = (dispatch: any): any =>
  bindActionCreators({ editToDo, deleteToDo }, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Challenge35);
