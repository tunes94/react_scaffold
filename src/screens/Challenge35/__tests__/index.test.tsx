import React from "react";
import { mount } from "enzyme";
import Challenge35 from "..";

const buildTestableComponent = (component: JSX.Element) => mount(component);

describe("<Challenge3.5 />", () => {
  it("should renders screen", (): void => {
    const wrapper = buildTestableComponent(<Challenge35 />);
    expect(wrapper).toBeDefined();
  });
});
