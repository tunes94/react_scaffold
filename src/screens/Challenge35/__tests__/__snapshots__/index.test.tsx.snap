// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Challenge3.5 /> should match snapshot 1`] = `ReactWrapper {}`;
